lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -98,13 +111,130 @@ import qualified Data.Map.Merge.Strict as Map
 import qualified Data.Map.Strict as Map
 import qualified Data.Text as T
 
--- | Stake Pool Data fields fetched from the node via LSQ
+--
+-- Stake Pool Layer
+--
+
+data StakePoolLayer = StakePoolLayer
+    { knownPools
+        :: IO [PoolId]
+
+    -- | List pools based given the the amount of stake the user intends to
+    -- delegate, which affects the size of the rewards and the ranking of the
+    -- pools.
+    , listStakePools
+        :: Coin
+        -> ExceptT ErrNetworkUnavailable IO [Api.ApiStakePool]
+    }
+
+newStakePoolLayer
+    :: GenesisParameters
+    -> NetworkLayer IO (IO Shelley) b
+    -> DBLayer IO
+    -> StakePoolLayer
+newStakePoolLayer gp nl db = StakePoolLayer
+    { knownPools = _knownPools
+    , listStakePools = _listPools
+    }
+  where
+    _knownPools
+        :: IO [PoolId]
+    _knownPools = do
+        tip <- getTip
+        let dummyCoin = Coin 0
+        res <- runExceptT $ map fst . Map.toList
+            . combineLsqData <$> stakeDistribution nl tip dummyCoin
+        case res of
+            Right x -> return x
+            Left _e -> return []
+
+    _listPools
+        :: Coin
+        -> ExceptT ErrNetworkUnavailable IO [Api.ApiStakePool]
+    _listPools userStake = do
+            tip <- liftIO getTip
+            lsqData <- combineLsqData <$> stakeDistribution nl tip userStake
+            chainData <- liftIO $ readDBPoolData db
+            return
+              . sortOn (Down . (view (#metrics . #nonMyopicMemberRewards)))
+              . map snd
+              . Map.toList
+              $ combineDBAndLsqData lsqData chainData
+
+    -- Note: We shouldn't have to do this conversion.
+    el = getEpochLength gp
+    gh = getGenesisBlockHash gp
+    getTip = fmap (toPoint gh el) . liftIO $ unsafeRunExceptT $ currentNodeTip nl
+
+--
+-- Data Combination functions
+--
+--
+
+-- | Stake Pool data fields that we can fetch from the node over Local State
+-- Query.
 data PoolLsqMetrics = PoolLsqMetrics
     { nonMyopicMemberRewards :: Quantity "lovelace" Word64
     , relativeStake :: Percentage
     , saturation :: Double
     } deriving (Eq, Show, Generic)
 
+-- | Stake Pool data fields that we read from the DB.
+data PoolDBMetrics = PoolDBMetrics
+    { regCert :: PoolRegistrationCertificate
+    , nProducedBlocks :: Quantity "block" Word64
+    , metadata :: Maybe StakePoolMetadata
+    }
+
+-- | Top level combine-function that merges DB and LSQ data.
+combineDBAndLsqData
+    :: Map PoolId PoolLsqMetrics
+    -> Map PoolId PoolDBMetrics
+    -> Map PoolId Api.ApiStakePool
+combineDBAndLsqData =
+    Map.merge lsqButNoChain chainButNoLsq bothPresent
+  where
+    lsqButNoChain = traverseMissing $ \k lsq -> pure $ mkApiPool k lsq Nothing
+
+    -- TODO: Maybe we have simply missed a retirement certificate. Maybe we
+    -- should simply dropMissing here.
+    chainButNoLsq = traverseMissing $ \k _ ->
+        error $ "Pool exists in DB but not in LSQ: " <> show k
```

Comment:
Yes we should! We do not track retirement certificates yet, so we definitely need this to account for them, same as we did for the ITN. When there's a pool in our database, but not returned by the LSQ, we fully trust the node and discard the pool on our end :+1:

Reply:
Ok, will change now

Reply:
Fixed.